src/librepgp/stream-parse.c

```diff
@@ -568,114 +503,183 @@ encrypted_src_close(pgp_source_t *src)
-        return RNP_ERROR_BAD_FORMAT;
-    }
+    pgp_cipher_cfb_decrypt(&crypt, enchdr, enchdr, blsize + 2);
+    if ((enchdr[blsize] == enchdr[blsize - 2]) && (enchdr[blsize + 1] == enchdr[blsize - 1])) {
+        src_skip(param->pkt.readsrc, blsize + 2);
+        param->decrypt = crypt;
+        /* init mdc if it is here */
+        /* RFC 4880, 5.13: Unlike the Symmetrically Encrypted Data Packet, no special CFB
+         * resynchronization is done after encrypting this prefix data. */
+        if (!param->has_mdc) {
+            pgp_cipher_cfb_resync(&param->decrypt);
+        } else {
+            if (!pgp_hash_create(&param->mdc, PGP_HASH_SHA1)) {
+                pgp_cipher_finish(&crypt);
+                (void) fprintf(stderr, "%s: cannot create sha1 hash\n", __func__);
+                return false;
+            }
 
-    // symmetric algorithm
-    skey->alg = buf[1];
+            pgp_hash_add(&param->mdc, enchdr, blsize + 2);
+        }
 
-    // s2k
-    skey->s2k.specifier = buf[2];
-    skey->s2k.hash_alg = buf[3];
-    len -= 4;
+        return true;
+    } else {
+        return false;
+    }
+}
 
-    switch (skey->s2k.specifier) {
-    case PGP_S2KS_SIMPLE:
+static bool
+encrypted_try_key(pgp_source_t *src, pgp_pk_sesskey_pkt_t *sesskey, pgp_key_t *key)
+{
+    uint8_t           decbuf[PGP_MPINT_SIZE];
+    rnp_result_t      err;
+    size_t            declen;
+    size_t            keylen;
+    pgp_fingerprint_t fingerprint;
+    pgp_seckey_t *    seckey = &key->key.seckey;
+    pgp_symm_alg_t    salg;
+    unsigned          checksum = 0;
+    bool              res = false;
+    BIGNUM *          ecdh_p;
+
+    /* Decrypting session key value */
+
+    switch (seckey->pubkey.alg) {
+    case PGP_PKA_RSA:
+        declen = pgp_rsa_decrypt_pkcs1(decbuf,
+                                       sizeof(decbuf),
+                                       sesskey->params.rsa.m,
+                                       sesskey->params.rsa.mlen,
+                                       &seckey->key.rsa,
+                                       &seckey->pubkey.key.rsa);
+        if (declen <= 0) {
+            (void) fprintf(stderr, "%s: RSA decryption failure\n", __func__);
+            return false;
+        }
         break;
-    case PGP_S2KS_SALTED:
-    case PGP_S2KS_ITERATED_AND_SALTED:
-        // salt
-        if (len < PGP_SALT_SIZE) {
-            return RNP_ERROR_BAD_FORMAT;
+    case PGP_PKA_SM2:
+        declen = sizeof(decbuf);
+        err = pgp_sm2_decrypt(decbuf,
+                              &declen,
+                              sesskey->params.sm2.m,
+                              sesskey->params.sm2.mlen,
+                              &seckey->key.ecc,
+                              &seckey->pubkey.key.ecc);
+
+        if (err != RNP_SUCCESS) {
+            (void) fprintf(
+              stderr, "%s: SM2 decryption failure, error %x\n", __func__, (int) err);
```

Comment:
``RNP_LOG`` please (or at least same output)

Reply:
Agree

Reply:
Okay, will update the PR